jvb/src/main/kotlin/org/jitsi/videobridge/relay/RelayMessageTransport.kt

```diff
@@ -0,0 +1,426 @@
+    parentLogger: Logger
+) : AbstractEndpointMessageTransport(parentLogger), ColibriWebSocket.EventHandler {
+    /**
+     * The last connected/accepted web-socket by this instance, if any.
+     */
+    private var webSocket: ColibriWebSocket? = null
+
+    /**
+     * For active websockets, the URL that was connected to.
+     */
+    private var url: String? = null
+
+    /**
+     * An active websocket client.
+     */
+    private var outgoingWebsocket: WebSocketClient? = null
+
+    /**
+     * Use to synchronize access to [.webSocket]
+     */
+    private val webSocketSyncRoot = Any()
+    private val numOutgoingMessagesDropped = AtomicInteger(0)
+
+    /**
+     * The number of sent message by type.
+     */
+    private val sentMessagesCounts: MutableMap<String, AtomicLong> = ConcurrentHashMap()
+
+    init { logger.addContext("relay-id", relay.id) }
+
+    /**
+     * Connect the bridge channel message to the websocket URL specified
+     */
+    fun connectTo(url: String) {
+        if (this.url != null && this.url == url) {
+            return
+        }
+        this.url = url
+
+        doConnect()
+    }
+
+    private fun doConnect() {
+        val url = this.url ?: throw IllegalStateException("Cannot connect Relay transport when no URL set")
+
+        webSocket = ColibriWebSocket(relay.id, this)
+        outgoingWebsocket = WebSocketClient().also {
+            it.start()
+            it.connect(webSocket, URI(url), ClientUpgradeRequest())
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun notifyTransportChannelConnected() {
+        relay.relayMessageTransportConnected()
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun clientHello(message: ClientHelloMessage): BridgeChannelMessage? {
+        // ClientHello was introduced for functional testing purposes. It
+        // triggers a ServerHello response from Videobridge. The exchange
+        // reveals (to the peer) that the transport channel between the
+        // remote relay and the Videobridge is operational.
+        // We take care to send the reply using the same transport channel on
+        // which we received the request..
+        return createServerHello()
+    }
+
+    override fun serverHello(message: ServerHelloMessage): BridgeChannelMessage? {
+        logger.info { "Received ServerHelloMessage, version ${message.version}" }
+        return null
+    }
+
+    /**
+     * This message indicates that a remote bridge wishes to receive video
+     * with certain constraints for a specific endpoint.
+     * @param message
+     * @return
+     */
+    override fun addReceiver(message: AddReceiverMessage): BridgeChannelMessage? {
+        val epId = message.endpointId
+        val ep = relay.conference.getEndpoint(epId) ?: run {
+            logger.warn("Received AddReceiverMessage for unknown epId $epId")
+            return null
+        }
+
+        ep.addReceiver(message.bridgeId, message.videoConstraints)
```

Comment:
I think we should use `relay.id` here instead of relying on the value in the message (the value in the message is redundant for relays). This way we're sure we can remove the correct receiver when the relay is expired.